enable single precision build for JCSDA-internal/MPAS-Model

metdyn opened the pull request:
## Description

This PR accompanies JCSDA-internal/mpas-jedi#728. 
It solves single precision build for MPAS-Model: issue #26

## To build
source env-setup/gnu-openmpi-cheyenne.sh 
set(MPAS_DOUBLE_PRECISION "OFF" CACHE STRING ....) in mpas-bundle/CMakeLists.txt
ecbuild (cmake); make

## Modified files
M       CMakeLists.txt
M       Makefile
M       cmake/PackageConfig.cmake.in

Commit:
create feature/sp

Commit:
M: Makefile
+       "FFLAGS_OPT = -O3 -m64 -ffree-line-length-none -fconvert=big-endian -ffree-form -fallow-argument-mismatch -fallow-invalid-boz"

Commit:
modify OpenMP in CMakeLists.txt

Commit:
mod

Commit:
This version works for single precision build

Commit:
modified:   Makefile

Commit:
delete an empty line

jjguerrette commented on Makefile:
```diff
@@ -294,7 +294,7 @@ gfortran:
 	"CXX_SERIAL = g++" \
 	"FFLAGS_PROMOTION = -fdefault-real-8 -fdefault-double-8" \
-	"FFLAGS_OPT = -O3 -m64 -ffree-line-length-none -fconvert=big-endian -ffree-form" \
+	"FFLAGS_OPT = -O3 -m64 -ffree-line-length-none -fconvert=big-endian -ffree-form -fallow-argument-mismatch -fallow-invalid-boz" \
```
It might be good to ask @mgduda what the implications of are for adding these two flags.  @metdyn, do we already use these flags in the cmake build?

mgduda replied:
For what it's worth, the current MPAS v7.3 release doesn't require the `-fallow-argument-mismatch -fallow-invalid-boz` flags when building with the GNU 9, 10, and 11 compilers. See, e.g., PR MPAS-Dev/MPAS-Model#736 .

jjguerrette replied:
Thank you Michael!  @metdyn, could you apply that same change in this branch, then remove the change to Makefile?  Eventually we need to merge MPASv7.3 into our fork(s).

metdyn replied:
@mgduda:  that is a good fix:
      integer, parameter:: bigint = int(O'17777777777')   
I will make corresponding changes to this set of code deposit.  Thanks for letting us know!

metdyn commented:
@mgduda:

An explicit question:
Is there a way to replace or modify `iso_c_binding` to alias or manipulate the keyword `double` in C++?
https://rgoswami.me/posts/iso-c-type-bound-fortran/

General description:
In C++, is there a way to redefine the Arithmetic Types: integral types and floating-point types:   For example: standard  int  (meaning integer) [size 16 bits];  float (meaning single-precision floating-point) [minimum size 6 significant digits]; double (meaning double-precision floating-point) [minimum size 10 significant digits].   For exmaple, re-define  in c++ code, 
```
  double frac = 0.0; 
  const double zz = sqrt(dot_product(gv, gv));
  std::vector<double>::iterator &   ...  ;
```
so that `double` becomes an alias to `float` single-precision. 
I feel there could be a correspondence in C or C++, to the FORTRAN functionality.

mgduda commented:
@metdyn I'm not sure that I understand your last question. Are you interested in C-Fortran interoperability of Fortran type-bound procedures? Or is the question how to be able to switch between `float` and `double` for interoperable variables?

metdyn commented:
> @metdyn I'm not sure that I understand your last question. Are you interested in C-Fortran interoperability of Fortran type-bound procedures? Or is the question how to be able to switch between `float` and `double` for interoperable variables?

I intend to ask
-  Is there a possibility to overwrite the `double` datatype in C or C++ code by `float`, so that the full C/C++ code alone (without considering any Fortran component) can execute in single precision mode?

Yes, I am interested in C-Fortran interoperability of Fortran type-bound procedures.
No. the question is how to switch `float` and `double` in the entire C code.

metdyn commented:
> > @metdyn I'm not sure that I understand your last question. Are you interested in C-Fortran interoperability of Fortran type-bound procedures? Or is the question how to be able to switch between `float` and `double` for interoperable variables?
> 
> I intend to ask
> 
> * Is there a possibility to overwrite the `double` datatype in C or C++ code by `float`, so that the full C/C++ code alone (without considering any Fortran component) can execute in single precision mode?
> 
> Yes, I am interested in C-Fortran interoperability of Fortran type-bound procedures. No. the question is how to switch `float` and `double` in the entire C code.

My error in using the word switch.  It is about either turn off `double` entirely or re-define (alias) `double` by `float` in C.

Commit:
Merge branch 'release-stable' into feature/sp

Commit:
create a duplicate branch from feature/sp

Commit:
o  module_ra_cam_support.F :  integer, parameter:: bigint = int(O'17777777777')
   doesn't require the -fallow-argument-mismatch -fallow-invalid-boz flags when building with
   the GNU 9, 10, and 11 compilers.
o  Remove the flags in Makefile

liujake reviewed (approved):
Some of PR message looks outdated with recent success of mpas-jedi single precision build. Need to update.

metdyn commented:
@liujake  You are right. I am updating the PR message.

jjguerrette reviewed (approved):
Great work @metdyn!

metdyn commented:
Will you check if I am on the authorized user list to merge?  (Merging is blocked)

jjguerrette commented:
@metdyn, I just added you with write privileges.

metdyn commented:
@jjguerrette  Thanks, but seems that something is still missing -- merging is blocked.

jjguerrette commented:
That's strange.  People with write access to mpas-jedi can merge, so I assumed it would be the same here.  Did you try closing your browser and re-opening?  Maybe deleting cookies?

jjguerrette commented:
I think I figured out what is different with mpas-jedi.  I just added the [JCSDA-internal/mpas](https://github.com/orgs/JCSDA-internal/teams/mpas) team as having PR merge ability.

metdyn commented:
@jjguerrette   I have restarted the mac.  The same message appears: 
Merging is blocked
The base branch restricts merging to authorized users. [Learn more about protected branches.](https://docs.github.com/articles/about-protected-branches/)

liujake commented:
I just merged it for you.

metdyn commented:
That works, thanks!

jjguerrette commented:
@metdyn, I think the problem was that I forget to click the "Save Changes" button.  I just made the same change and clicked the button.  Hopefully it works for you and others next time.

metdyn commented:
@JJ, good you figured it out. It happens and hope it’s all good now!  

> On May 27, 2022, at 1:20 PM, JJ Guerrette ***@***.***> wrote:
> 
> ﻿
> @metdyn, I think the problem was that I forget to click the "Save Changes" button. I just made the same change and clicked the button. Hopefully it works for you and others next time.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.
